### PR TITLE
fix: app crash on scrolling in skills

### LINKED ISSUE #1647: App crashes on scrolling on skills activity
**Actual Behaviour**

Currently the app crashes after scrolling down in the skills activity because the URLs with space are not getting parsed.

**Expected Behaviour**

The app should not crash

**Steps to reproduce it**

Open skills activity and scroll down.

**Would you like to work on the issue?**

Yes

### PR DESCRIPTION
**Pull request by batbrain7**
Fixes #1647

### CODE REVIEW

File: app/src/main/java/org/fossasia/susi/ai/helper/Utils.kt [hunk 1]
```diff
@@ -18,8 +19,10 @@ object Utils {
     }
 
     fun getImageLink(skillData: SkillData): String {
-        val uri = URI("${BaseUrl.SUSI_DEFAULT_BASE_URL}/cms/getImage.png?model=${skillData.model}&language=${skillData.language}&group=${skillData.group}&image=${skillData.image}")
-        return uri.toASCIIString()
+        val s = "${BaseUrl.SUSI_DEFAULT_BASE_URL}/cms/getImage.png?model=${skillData.model}&language=${skillData.language}&group=${skillData.group}&image=${skillData.image}"
+                .replace(" ","%20")
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/helper/Utils.kt]: Obviously, you are not going to release new app right away. Add tests now as previously instructed. In order to save time in that PR, you are going to spend it here 

Also, I told you that save the single letter variable names for your college projects

**Comment by batbrain7** [comment]
@iamareebjamal tests show empty test suite and class not found exception, I am unable to set the configuration?

**Comment by arundhati24** [comment]
@iamareebjamal I don't understand why the app didn't crash for me. I could see the top games in the skills listing activity. This is a valid issue that spaces should be replaced with '%20' but I don't really get why it doesn't crash when I run it on my device before sending PRs and the app starts crashing immediately after release. This has happened twice or thrice. Is this a problem with instant run or do you have an idea what could be the possible cause of this?

**Comment by batbrain7** [comment]
@arundhati24 same, it worked fine for me in the debug version

**Comment by arundhati24** [comment]
@batbrain7 Yes. I test the app after every single commit. And, this is strange that it didn't crash before releasing. :(

File: app/src/androidTest/java/org/fossasia/susi/ai/helper/ValidateUtilsTest.kt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        assertFalse(ValidateUtils.validateUrl("http://."))
+        assertFalse(ValidateUtils.validateUrl("http://...."))
+        assertFalse(ValidateUtils.validateUrl("http56://www.test.com"))
+        assertFalse(ValidateUtils.validateUrl("http//www.test."))
+        assertFalse(ValidateUtils.validateUrl("http//www."))
+        assertFalse(ValidateUtils.validateUrl("httpstest.com"))
+        assertFalse(ValidateUtils.validateUrl("test"))
+        assertFalse(ValidateUtils.validateUrl("test.co.in"))
+        assertFalse(ValidateUtils.validateUrl(".com"))
+    }
```
Comment (iamareebjamal) [inline app/src/androidTest/java/org/fossasia/susi/ai/helper/ValidateUtilsTest.kt]: Useless tests. You should test the method I commented on and not a new method you created which isn't used anywhere

**Comment by iamareebjamal** [comment]
This is maybe because you test the debug version and not the release version

File: app/src/androidTest/java/org/fossasia/susi/ai/helper/ValidateUtilsTest.kt (hunk 2 above)
Comment (batbrain7) [reply app/src/androidTest/java/org/fossasia/susi/ai/helper/ValidateUtilsTest.kt]: Do I need to mock the skildata object?

Comment (iamareebjamal) [reply app/src/androidTest/java/org/fossasia/susi/ai/helper/ValidateUtilsTest.kt]: No, you need to create it

Comment (batbrain7) [reply app/src/androidTest/java/org/fossasia/susi/ai/helper/ValidateUtilsTest.kt]: @iamareebjamal I am unable to understand how to create the test, could you help a little ?

Comment (batbrain7) [reply app/src/androidTest/java/org/fossasia/susi/ai/helper/ValidateUtilsTest.kt]: @arundhati24 it would be great if you could also help here a bit, as a new release has to be after this on both FDroid and Playstore

Comment (iamareebjamal) [reply app/src/androidTest/java/org/fossasia/susi/ai/helper/ValidateUtilsTest.kt]: Pass in SkillData object and verify the URL which will return from the function

File: app/src/androidTest/java/org/fossasia/susi/ai/helper/UtilsTest.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import android.util.Patterns
+import junit.framework.Assert.assertTrue
+import org.fossasia.susi.ai.rest.responses.susi.SkillData
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class UtilsTest {
+
```
Comment (batbrain7) [inline app/src/androidTest/java/org/fossasia/susi/ai/helper/UtilsTest.kt]: @iamareebjamal since there are no Unit tests in the app right now , on running the tests ClassNotFound is shown each time, even running an empty test gives this error.

File: app/build.gradle [hunk 4]
```diff
@@ -110,11 +117,14 @@ dependencies {
...
     //unit test
+    androidTestImplementation('com.android.support.test.espresso:espresso-core:3.0.1', {
+        exclude group: 'com.android.support', module: 'support-annotations'
+    })
     androidTestImplementation "com.android.support:support-annotations:${rootConfiguration.supportLibraryVersion}"
     androidTestImplementation "com.android.support.test:runner:${rootConfiguration.testVersion}"
     androidTestImplementation "com.android.support.test:rules:${rootConfiguration.testVersion}"
     androidTestImplementation "com.android.support.test.espresso:espresso-core:${rootConfiguration.espressoVersion}"
-    testImplementation 'junit:junit:4.12'
+    androidTestImplementation 'junit:junit:4.12'
```
Comment (iamareebjamal) [inline app/build.gradle]: Revert

File: app/src/androidTest/java/org/fossasia/susi/ai/helper/UtilsTest.kt (hunk 3 above)
Comment (iamareebjamal) [reply app/src/androidTest/java/org/fossasia/susi/ai/helper/UtilsTest.kt]: Uhh. This is androidTest. Please add in the unit test directory

File: app/src/androidTest/java/org/fossasia/susi/ai/chat/ChatActivityTest.kt [hunk 5]
```diff
@@ -74,10 +74,10 @@ class ChatActivityTest {
         onView(withId(R.id.rv_chat_feed)).check(matches(isDisplayed()))
 
         // checks if layout container for chat box is present
-        onView(withId(R.id.send_message_layout)).check(matches(isDisplayed()))
+//        onView(withId(R.id.send_message_layout)).check(matches(isDisplayed()))
 
         // checks if message box is present
-        onView(withId(R.id.et_message)).check(matches(isDisplayed()))
+//        onView(withId(R.id.et_message)).check(matches(isDisplayed()))
```
Comment (iamareebjamal) [inline app/src/androidTest/java/org/fossasia/susi/ai/chat/ChatActivityTest.kt]: Revert

File: app/src/androidTest/java/org/fossasia/susi/ai/chat/ChatActivityTest.kt [hunk 6]
```diff
@@ -74,10 +74,10 @@ class ChatActivityTest {
         onView(withId(R.id.rv_chat_feed)).check(matches(isDisplayed()))
 
         // checks if layout container for chat box is present
-        onView(withId(R.id.send_message_layout)).check(matches(isDisplayed()))
+//        onView(withId(R.id.send_message_layout)).check(matches(isDisplayed()))
```
Comment (iamareebjamal) [inline app/src/androidTest/java/org/fossasia/susi/ai/chat/ChatActivityTest.kt]: Revert

**Comment by iamareebjamal** [comment]
Please revert all the changes. There should be only 2 files changed. One the buggy file and other one the test

**Comment by iamareebjamal** [comment]
If you looked at this PR https://github.com/fossasia/susi_android/pull/1620

You would have known how to implement tests properly and would have saved a few hours

File: app/src/test/java/org/fossasia/susi/ai/helper/helper/UtilsTest.kt [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+import android.util.Patterns
+import junit.framework.Assert.assertTrue
+import org.fossasia.susi.ai.helper.Utils
+import org.fossasia.susi.ai.rest.responses.susi.SkillData
+import org.junit.Test
+
+class UtilsTest {
+
+    val skillData: SkillData = SkillData()
```
Comment (batbrain7) [inline app/src/test/java/org/fossasia/susi/ai/helper/helper/UtilsTest.kt]: @iamareebjamal since the SkillData is always parcelable it returns null pointer exception

File: app/src/test/java/org/fossasia/susi/ai/helper/UtilsTest.kt [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+import org.fossasia.susi.ai.rest.responses.susi.SkillData
+import org.junit.Test
+
+class UtilsTest {
+
+    @Test
+    fun verifyImageLink() {
+        val str = Utils.getImageLink()
+        val yes = ValidateUtils.validateUrl(str)
+        assertTrue(yes)
```
Comment (iamareebjamal) [inline app/src/test/java/org/fossasia/susi/ai/helper/UtilsTest.kt]: Wrong

Comment (iamareebjamal) [reply app/src/test/java/org/fossasia/susi/ai/helper/UtilsTest.kt]: First right the tests, then fix it

**Comment by iamareebjamal** [comment]
@batbrain7 In how many days will this be finalized?

**Comment by batbrain7** [comment]
@iamareebjamal I do not understand what to do next, the regex does not detect spaces

File: app/src/test/java/org/fossasia/susi/ai/helper/UtilsTest.kt [hunk 9]
```diff
@@ -1,15 +1,23 @@
...
-        val str = Utils.getImageLink()
-        val yes = ValidateUtils.validateUrl(str)
-        assertTrue(yes)
+        assertEquals("https://api.susi.ai/cms/getImage.png?model=model&language=test&group=abc%20def%20ghi&image=images/test.png",
+                Utils.getImageLink(SkillData(model = "model", group = "abc def ghi", language = "test", image = "images/test.png")))
+
+        assertEquals("https://api.susi.ai/cms/getImage.png?model=model&language=test&group=abcdefghi&image=images/test.png",
+                Utils.getImageLink(SkillData(model = "model", group = "abcdefghi", language = "test", image = "images/test.png")))
+
+        assertEquals("https://api.susi.ai/cms/getImage.png?model=model&language=test&group=abc%20def%20ghi&image=images/test_image.jpg",
```
[the remark this answers is not in the document]
Comment (batbrain7) [reply app/src/test/java/org/fossasia/susi/ai/helper/UtilsTest.kt]: No @iamareebjamal the image string is different

[the remark this answers is not in the document]
Comment (iamareebjamal) [reply app/src/test/java/org/fossasia/susi/ai/helper/UtilsTest.kt]: String might be different, the test is not testing anything extra than the first one. Tests should only be aded to check edge cases and separate branches. The two tests above are enough, but you can add more tests if they are testing new things

[the remark this answers is not in the document]
Comment (batbrain7) [reply app/src/test/java/org/fossasia/susi/ai/helper/UtilsTest.kt]: Updating the PR asap

**Comment by batbrain7** [comment]
after this PR we update the crashing version that is on the playstore, @iamareebjamal is this PR good to go ?

File: app/src/main/java/org/fossasia/susi/ai/helper/Utils.kt [hunk 10]
```diff
@@ -24,6 +23,13 @@ object Utils {
                 .into(imageView)
     }
 
+    fun getImageLink(skillData: SkillData): String {
+        val link = "${BaseUrl.SUSI_DEFAULT_BASE_URL}/cms/getImage.png?model=${skillData.model}&language=${skillData.language}&group=${skillData.group}&image=${skillData.image}"
+                .replace(" ", "%20")
+        Timber.d("SUSI URI" + link)
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/helper/Utils.kt]: Pass as argument and don't append

**Comment by batbrain7** [comment]
@iamareebjamal on passing as argument android studio displayed an error and asked to add Suppress lint, so instead I used kotlin string interpolation

File: app/src/main/java/org/fossasia/susi/ai/helper/Utils.kt [hunk 11]
```diff
@@ -24,6 +24,13 @@ object Utils {
                 .into(imageView)
     }
 
+    fun getImageLink(skillData: SkillData): String {
+        val link = "${BaseUrl.SUSI_DEFAULT_BASE_URL}/cms/getImage.png?model=${skillData.model}&language=${skillData.language}&group=${skillData.group}&image=${skillData.image}"
+                .replace(" ", "%20")
+        Timber.d("SUSI URI ${link}")
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/helper/Utils.kt]: Still not passed as argument
